### PR TITLE
Feature/77 gdpr remove data

### PR DESCRIPTION
**Pull request by wotwot563**
## Description

This pull request adds the feature of deleting userdata so that dex is gdpr compliant.

## Type of change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] This change requires a documentation update

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I did not update API Controllers, if I did, I added/changed Postman tests
-   [x] I have performed a self-review of my own code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [x] I have made corresponding changes to the documentation
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [x] New and existing unit tests pass locally with my changes
-   [x] I updated the changelog with an end-user readable description
-   [x] I assigned this pull request to the correct project board to update the sprint board

## Steps to Test or Reproduce
1. Re-seed your database so the placeholder user is created.
2. Create a user.
2. Add a project as that user.
3. Embed the created project
4.  make a DELETE request to the /api/User/{userid}
5. Now if you check the database:
    - The project table entries should be updated, the userId should be 2 (placeholder user)
    - The user table entry should be deleted.
    - The identity user entry should be deleted as well.

## Link to issue

Closes: #77

### CODE REVIEW

**Review by MeesvanStraten — CHANGES_REQUESTED** [review]
After replicating the steps you listed the table "Embedded project" has no record of the deleted user. (Thats good i think). The project however does still exist in the "Project" table and has the userId set to 1 is that how it should be? Also i think userId 1 is admin Bob for me. The Removed User i do not see anywhere.

**Comment by niraymak** [comment]
Just executed steps to reproduce. Used my FHICT account to create a new account.
Then added and embedded a project. 

Continued to step 4. Response without identityId was internal server error. prIdentityId and registeredIdentityId returned 401 👍 
https://gyazo.com/3efe339f6b705ae983c162c84e82e0af
Maybe you should look into this.

After step 4 my user record and (embed)project records are deleted.
I also did the previous steps and then did step 5 instead of step 4. After this my data was not deleted. Is it supposed to work like this? What is the IdentityUrl/Account/Delete page for if it doesn't do anything?

Nice work!

**Review by niraymak — APPROVED** [review]
Nice code, works like it is supposed to work. 👍 

For reviewers: Was a bit unclear for me that the projects are not deleted but taken over by a new "placeholder" user. This "placeholder" reviewer is created in the seed. So make sure to run the seed (delete database) before testing.

**Review by niraymak — CHANGES_REQUESTED** [review]
Wondering though, how would this work in production. 
Is it required for both admin (or yourself) to use delete endpoint /api/User/{userId} and delete yourself in the identityId? Wouldn't it be easier to do it by one click?

Also postman tests don't seem to be passing.

**Comment by wotwot563** [comment]
> 
> 
> Wondering though, how would this work in production.
> Is it required for both admin (or yourself) to use delete endpoint /api/User/{userId} and delete yourself in the identityId? Wouldn't it be easier to do it by one click?
> 
> Also postman tests don't seem to be passing.

I thought about the process for production and came up with the following:

1. user requests gdpr data removal
2. data is deleted in the backend.
3. user is redirected to the identity server.
4. user approves data deletion again (just to make sure)
5. idsrv data is deleted
6. user is redirected back to the frontend with a generic success page.

for admin you would have to create a crud ui admin page. and in the identity server too, but because its an mvc pattern its easier to just call the user delete method with a list of users to keep the requests on the low side.

the code responsible for deleting users in the idsrv is:
```
IdentityUser user = await identityUserService.FindAsync(sub);
if(user != null)
{
  identityUserService.Remove(user);
  identityUserService.Save();
}
```

For the postman tests hadnt had the time to fix them yet

**Comment by MeesvanStraten** [comment]
Saw this PR was still open, don't know what the status is so just responding to the latest comment here 😄 

> 4\. user approves data deletion again (just to make sure)

Would it not be easier if there was 1 end point responsible to call from the front end that handles everything? So that end point could call the identity server and all other necessary services to delete all data from the user.
Then if all succeeds the endpoint returns 200 and the front end knows all data has been wiped.

Cheers

**Comment by Brend-Smits** [comment]
@waltersajtos do you want to pick this up and finish it?

**Comment by waltersajtos** [comment]
I moved everything into 1 controller because I could not find a reason to separate it. I also added the tests so I think this can be merged after the tests have been refactored by @Ruby77.

Can you check if you agree with me @Brend-Smits?

**Comment by niraymak** [comment]
Record in EmbeddedProject is being deleted. Is it supposed to work like that?

File: API/Controllers/UserController.cs
Comment (niraymak) [inline API/Controllers/UserController.cs:321]: Got a status 500 when not passing an userId. "Nullable object must have a value."

Comment (niraymak) [reply API/Controllers/UserController.cs:321]: Just saw someone did use the same GUID in this controller on line: 423, 433. Can you please fix it :)

**Review by niraymak — CHANGES_REQUESTED** [review]
Tried to delete without userId but got a error code 500 and test don't succeed yet.
Noticed a duplicate GUID which got through, can you fix it? Thanks in advance!

Keep up the good work!

**Comment by Brend-Smits** [comment]
I will review this once @niraymak approves it. Ping me when this happens.

**Comment by waltersajtos** [comment]
> Record in EmbeddedProject is being deleted. Is it supposed to work like that?

I think so yeah, I didn't think out the logic. Just made work what was already there

**Comment by waltersajtos** [comment]
I think this pull request is finished now, please double check the steps to reproduce. I just updated them

File: IdentityServer/Quickstart/Account/AccountController.cs
Comment (MeesvanStraten) [inline IdentityServer/Quickstart/Account/AccountController.cs:292]: Async method is not doing anything I think? might remove it.

**Review by MeesvanStraten — DISMISSED** [review]

File: IdentityServer/Quickstart/Account/AccountController.cs
Comment (waltersajtos) [reply IdentityServer/Quickstart/Account/AccountController.cs:292]: I will remove the async, the redirect is not necessary anymore since the identity user and application user will be deleted at the same time now.

**Review by MeesvanStraten — APPROVED** [review]

**Comment by niraymak** [comment]
In Docker environment there seem to be a few problems. I mentioned you added a dependency on IdentityUserService for the API. I think only the IdentityServer should have a dependency on it, because it's the IdentityServer's responsiblity. I think the way to go is to make a REST call to the IdentityServer. But I'd like to hear someone else's opinion on this.

**Comment by MeesvanStraten** [comment]
I agree with Niray on having no dependency on the IdentityUserService. My opinion would be to as Niray suggested make a http call to the IdentityServer.

**Comment by waltersajtos** [comment]
There currently is no identity server end-point to delete a user trough a rest request, we need to decide if we want to implement this or just redirect the user like wotwot proposed.

@niraymak @Brend-Smits

**Comment by niraymak** [comment]
I guess create new request in IdentityServer.
Call the request in the API request. If it returns 200 then delete GDPR data.

**Comment by waltersajtos** [comment]
I'm not sure if it's a good idea to add an end point to the identity server to be honest, there is no way to check whether the user is allowed to delete the account or not. I don't think we want to rush this one, it's quite essential that this is secure.

**Comment by niraymak** [comment]
Do you want to continue with this issue? @waltersajtos

**Comment by waltersajtos** [comment]
> Do you want to continue with this issue? @waltersajtos 

If anyone wants to pick this up, please do it. I guess it might be better to start from scratch tbh. The problem I ran into with the current implementation is deleting the data from the identity server in a safe way.

**Pull request closed without merging**
